### PR TITLE
HyP-183 Offline compatible

### PR DESCRIPTION
**Pull request by dc-andysign**
---
name: Offline Compatible
about: Make demonstrator available when offline
---

## Summary

This PR is in connection to the **Jira** ticket **HYP-183**.

At the moment we must be online for the demonstrator. The person should be able to connect to the demonstrator without internet (such as the demonstrator provides embodied CO2 even when the API is not connected.)

Use case: For example conference purposes.

---

## Motivation

To allow for the demo to be performed in all conditions.

---

## Describe alternatives you've considered

In terms of alternatives, I can list the following:

1. Change in the API, as in, the addition of a fall back block of code to be execute when off-line mode detected **`src/lib/services/emissionsCalculator.ts`**. This has been abandoned as the consensus seems to be to refrain from changing the API at this point.

2. Copy over **React** and **ReactDom** from the modules folder and change the index.html to point to those files instead of the online ones. This has been scraped because it doesn't seem to be an elegant solution ( having 2 identical files in 2 places is not optimal ).

3. Copy over **React** and **ReactDom** from the modules folder and change the index.html and add a fallback into the index html with **`<script "https://unpkg.com/file" onerror="importLocal()"></script>`**. This has been abandoned as it doesn't seem to work.

4. Integrating the webpack **offline-plugin**. This is better described **[here](https://github.com/kay-is/webpack-offline-plugin-example/tree/master)** and **[here](https://www.npmjs.com/package/offline-plugin)**. This has been dropped as the NPM library seems to be inactive.

---

## Additional context

To make the GUI work offline the easies solution is to empty the **`externals: {}`** object from **`webpack.config.js`**. Adding things like **react** and **react-dom** in that object means preventing webpack from adding those items into the bundle and therefore relaying on the CDN.

screenshot

Pinned down to a specific version ( NOT **latest** version ) all dependencies to make sure all the containers will start even when off-line.

image

Changed the **initialise** script ( **`scripts/initialise_demo.mjs`** ) to check if the demo is in off-line mode and if yes to add custom random co2 values to the body of the POST call.

image

Changed the **GUI** so that the code checks if the demo is in off-line mode and if yes add custom random Co2 value to the issuance POST call.

image

Also changed the code to keep the init script under one minute:

image

---

### CODE REVIEW

File: src/index.html [hunk 1]
```diff
@@ -1,8 +1,6 @@
 <!DOCTYPE html>
 <html lang='en'>
   <head>
-    <script crossorigin src="https://unpkg.com/react@18/umd/react.production.min.js"></script>
-    <script crossorigin src="https://unpkg.com/react-dom@18/umd/react-dom.production.min.js"></script>
     <script src='/config.js' type='module'></script>
```
Comment (mattdean-digicatapult) [inline src/index.html:4]: Is react being bundled in from `node_modules` now presumably?

File: webpack.config.js [hunk 2]
```diff
@@ -74,9 +74,9 @@ module.exports = (vars) => {
         analyzerMode: 'disabled', // disable analyser. Replace with "server" or "static" to render analyser
       }),
     ],
-    externals: {
-      react: 'React',
-      'react-dom': 'ReactDOM',
+    externals: { // externals object should be empty to include React & ReactDOM in the bundle making the GUI available offline
+      // react: 'React',
+      // 'react-dom': 'ReactDOM',
```
Comment (mattdean-digicatapult) [inline webpack.config.js:79]: Can we just remove this?

**Review by mattdean-digicatapult — COMMENTED** [review]
I think this looks good. Before you merge though I'd like to disable renovate on this repository otherwise all the version pinning will leave us with tonnes of PR notifications. If you're doing that please make sure this repo is marked as non-production software

File: docker/hyproof-apis.yml [hunk 3]
```diff
@@ -19,11 +19,11 @@ services:
 
   hydrogen-producer-hyproof-api:
     container_name: hydrogen-producer-hyproof-api
-    image: digicatapult/sqnc-hyproof-api:latest
+    image: digicatapult/sqnc-hyproof-api:v0.13.25
```
Comment (rmlearney-digicatapult) [inline docker/hyproof-apis.yml:22]: check all current versions just before final merge

File: src/index.html (hunk 1 above)
Comment (dc-andysign) [reply src/index.html:4]: Yes, indeed, this means _react_ and _react-dom_ will be included there in the bundle, e.g. : 

```
> ls
bundle.0dca980c8297507f0608.js
react-dom.production.min.js
react.production.min.js
```

I'm happy for this solution as it is really nice and _elegant_ ..

File: webpack.config.js (hunk 2 above)
Comment (dc-andysign) [reply webpack.config.js]: No problem. DONE.

File: docker/hyproof-apis.yml (hunk 3 above)
Comment (dc-andysign) [reply docker/hyproof-apis.yml:22]: DONE.

Comment (dc-andysign) [reply docker/hyproof-apis.yml]: Hmmm , actually I need to deactivate renovate first which is this PR here:

https://github.com/digicatapult/sqnc-hyproof-client/pull/148

**Review by mattdean-digicatapult — APPROVED** [review]

**Review by dblane-digicatapult — APPROVED** [review]

File: scripts/initialise_demo.mjs [hunk 4]
```diff
@@ -174,6 +174,50 @@ async function initiateCertificate(hydrogenProducerPort, energyProviderPort) {
   return waitForComplete
 }
 
+// async function checkCarbonIntensityAPI(url, timeout = 2000) {
```
Comment (rmlearney-digicatapult) [inline scripts/initialise_demo.mjs:177]: Don't comment, just delete

Comment (dblane-digicatapult) [reply scripts/initialise_demo.mjs:177]: With the stuff that has been commented here, I'm not sure that is the correct solution.

Comment (dc-andysign) [reply scripts/initialise_demo.mjs]: All cleaned !

File: docker/hyproof-apis.yml (hunk 3 above)
Comment (dc-andysign) [reply docker/hyproof-apis.yml]: > check all current versions just before final merge

Just finished doing another check ..

**Review by rmlearney-digicatapult — APPROVED** [review]

**Pull request merged**